### ISSUE TITLE
switch to rails_config 0.5.0

### ISSUE THREAD

**Issue by tillmo** [ontohub/ontohub#1231]
@jelmd needs this for his work on settings. However, switching to rails_config 0.5.0 seems to cause problems in lib/environment_light.rb. Can this be fixed?

**Comment by jelmd** [comment]
It's actually already part of https://github.com/ontohub/ontohub/pull/1206/files

**Comment by tillmo** [comment]
not really, because #1206 breaks the git clone functionality. We should have tests for this...

**Comment by jelmd** [comment]
Can you give the line numbers, which cause the break? Should be easy to fix than ...

**Comment by tillmo** [comment]
It is the stuff that we discussed on Wednesday afternoon. I do not have access to td any longer, so I cannot tell details.

**Comment by jelmd** [comment]
Access should be fixed - authorized keys were missing.

**Comment by tillmo** [comment]
Here is the output that I get (different from Wednesday's output):
```
till@MARK ~/temp> git clone git@td.iws.cs.ovgu.de:test.git
Klone nach 'test'...
Warning: the ECDSA host key for 'td.iws.cs.ovgu.de' differs from the key for the IP address '141.44.24.83'
Offending key for IP in /home/till/.ssh/known_hosts:363
Matching host key in /home/till/.ssh/known_hosts:364
Are you sure you want to continue connecting (yes/no)? yes
/local/home/ontohub/ontohub/git/bin/git-shell: line 6: unless: command not found
/local/home/ontohub/ontohub/git/bin/git-shell: line 7: STDERR.puts: command not found
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
128 till@MARK ~/temp> 
```

**Comment by jelmd** [comment]
fixed - she bang error in the zone setup script.

**Comment by tillmo** [comment]
now I get:
```
till@MARK ~/temp> git clone git@td.iws.cs.ovgu.de:test.git
Klone nach 'test'...
Warning: the ECDSA host key for 'td.iws.cs.ovgu.de' differs from the key for the IP address '141.44.24.83'
Offending key for IP in /home/till/.ssh/known_hosts:363
Matching host key in /home/till/.ssh/known_hosts:364
Are you sure you want to continue connecting (yes/no)? yes
fatal: '/data/git/repositories/test.git' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
128 till@MARK ~/temp> 
```

**Comment by jelmd** [comment]
Can you remove the lines 363 and 364 from your .ssh/known_hosts, plz ?

**Comment by jelmd** [comment]
Sorry, missed  the fatal lines - yes, that is, what I meant with "mixup".

**Comment by tillmo** [comment]
OK, did that, but the error remains the same.

**Comment by tillmo** [comment]
it seems that the server transmits some text that does not follow the git protocol.

**Comment by jelmd** [comment]
Hah, couldn't resist to check this and found the reason, when looking @bremen instance: there git-shell and the web app use different data_dirs and thus different git_roots, i.e. `~git` (which is `/srv/http/ontohub/shared/data/git_user/`) vs. `/srv/http/ontohub/releases/data` (which in turn is a symlink to `/srv/http/ontohub/shared/data/`)! This works, because git-shell's git_root `~git/repositories` is a symlink to `/srv/http/ontohub/shared/data/git_daemon/`, which really contains the symbolic names of the repos (but the git_root of the web app is `data/repositories/`, which finally resolves to `/srv/http/ontohub/shared/data/repositories/`).

Because the  1170-git-data_dir.patch fixes these inconsistent behavior, the web app as well as the git-shell use the same settings and thus the same data_dir (which is for us `/data/git/`) and therefore the same deduced git_root: `/data/git/repositories/`, which hosts the repos, but under its ID. So now I'm pretty sure, that `s/git_root/symlink_path/g` in the related files below git/ is the correct fix.

Made the changes on td and it seems to work.
See also https://github.com/ontohub/ontohub/blob/staging/config/initializers/paths.rb

**Comment by eugenk** [comment]
With the referenced pull request, the current version of rails_config is used. It also enables array overwriting (see commit message of 22b0b434f6a92b32c4c9364c11d625f34af67dd9 for an example).

As far as I remember from Wednesday, this is something we need for the `hets_path` config to be secure.

**Comment by jelmd** [comment]
Not really, because using a different type actually replaces the previous setting and thus we already have, what we need wrt. to the settings itself. So just using a string instead of an array does the job, and we only need a single path. If this wouldn't be enough, i.e. wanna replace an array with an array, one could use the trick to go over 2 other files, one replaces it with a str, and the 2nd replaces it again with an array...

Anyway, all such tricks or pseudo fixes are not needed at all, if the app would actually support reasonable defaults (as shown in 1170-settings-cleanup.patch). 

The need for RailsConfig 0.5.0 is just what you probably would call "Don't repeat yourself" - me "encapsulation"/"don't re-invent the wheel" => keep its behavior consistent, i.e. always read the same file in the same order - see https://github.com/railsconfig/rails_config/blob/master/lib/rails_config.rb#L43 ...

**Comment by eugenk** [comment]
> using a different type actually replaces the previous setting

We don't want to use a different type because it makes the code that uses the setting mode complex. Since ontohub is a very complex application itself, we don't want to make the code even more complex by obscuring the usage of the settings.

>  i.e. wanna replace an array with an array, one could use the trick to go over 2 other files, one replaces it with a str

Also, going over two files is rather a dirty hack than a clean solution.

**Comment by jelmd** [comment]
wrt 1) Grrr, you got it wrong! It is what I told/show you on Wed. already: If there a structure in the yaml file for a certain setting causes RailsConfig to choose an array for this and in the next file one uses a structure, which implies or even forces a different type for this setting (e.g. str,) than the previous setting gets overwritten completely, otherwise, if both are compatible, they get merged. That's the default behavior. Wondering, whether you have studied the code before forking it - its explicitly mentioned in https://github.com/eugenk/rails_config/blob/master/lib/rails_config/vendor/deep_merge.rb !

wrt. 2) That's what you actually force with your latest patches. Not sure, what you mean with clean, but this is just a different incarnation of the environment* stuff - I don't like it either, but I got told, that's the usual rails way ... (and as said again and again, a smart application should not need a config file at all - however, it should allow one to tune it to its needs using ONE config file ;-), or have you ever seen, that someone spent an hour+ to configure its new smartphone app or even that one has to read a manual to be able to do that?

BTW: Wondering, whether you have ever taken a minute to look over the patches I've already submitted (most recent ones can always be found on td:~ontohub/etc/)? Why not just correct the things you don't like, apply and your are through with it and can start to deal with the interesting stuff again?
